File: gge/gge/math/Vec3.cpp
#include "Vec3.h"
#include "Quat.h"
#include <cassert>
#include <sstream>

namespace gge
{
    
    const Vec3 Vec3::UNIT_X(1,0,0);
    const Vec3 Vec3::UNIT_Y(0,1,0);
    const Vec3 Vec3::UNIT_Z(0,0,1);
    const Vec3 Vec3::ZERO;
    
    Vec3::Vec3() :
        x(0),
        y(0),
        z(0)
    {
    }
    
    Vec3::Vec3(float _x, float _y, float _z) :
        x(_x),
        y(_y),
        z(_z)
    {
    }
    
    Vec3::Vec3(const Quat& q, float length)
    {
        *this = (q * Vec3::UNIT_X) * length;
    }
    
    Vec3::Vec3(const Vec3& v) :
        x(v.x),
        y(v.y),
        z(v.z)
    {
    }
    
    Vec3& Vec3::operator=(const Vec3& v)
    {
        x = v.x;
        y = v.y;
        z = v.z;
        return *this;
    }
    
    float& Vec3::operator[](int i)
    {
        assert(i >= 0 && i < 3);
        if (i == 0) 
            return x;
        else if (i == 1)
            return y;
        else
            return z;
    }
    
    float Vec3::operator[](int i) const
    {
        assert(i >= 0 && i < 3);
        if (i == 0) 
            return x;
        else if (i == 1)
            return y;
        else
            return z;
    }
    
    Vec3 Vec3::operator+(const Vec3& v) const
    {
        return Vec3(x + v.x, y + v.y, z + v.z);
    }
    
    Vec3 Vec3::operator-(const Vec3& v) const
    {
        return Vec3(x - v.x, y - v.y, z - v.z);
    }
    
    Vec3 Vec3::operator*(float f) const
    {
        return Vec3(x * f, y * f, z * f);
    }
    
    Vec3 Vec3::operator/(float f) const
    {
        assert(f != 0);
        return Vec3(x / f, y / f, z / f);
    }
    
    Vec3 Vec3::operator-() const
    {
        return Vec3(-x,-y,-z);
    }
    
    void Vec3::operator+=(const Vec3& v)
    {
        x += v.x;
        y += v.y;
        z += v.z;
    }
    
    void Vec3::operator-=(const Vec3& v)
    {
        x -= v.x;
        y -= v.y;
        z -= v.z;
    }
    
    void Vec3::operator*=(float f)
    {
        x *= f;
        y *= f;
        z *= f;
    }
    
    void Vec3::operator/=(float f)
    {
        assert(f != 0);
        x /= f;
        y /= f;
        z /= f;
    }
    
    float Vec3::length() const
    {
        return Math::sqrt(x * x + y * y + z * z);
    }
    
    Vec3 Vec3::norm() const
    {
        return *this / length();
    }
    
    float Vec3::dot(const Vec3& v) const
    {
        return x * v.x + y * v.y + z * v.z;
    }
    
    Vec3 Vec3::cross(const Vec3& v) const
    {
        return Vec3(y * v.z - z * v.y,
                    z * v.x - x * v.z,
                    x * v.y - y * v.x);
    }
    
    Vec3 Vec3::project(const Vec3& v) const
    {
        float l = length();
        assert(l != 0);
        return *this * (dot(v) / (l * l));
    }
    
    Quat Vec3::getRotTo(const Vec3& v) const
    {
        float d = dot(v);
        if (d > 0.99999)
            return Quat();
        else if (d < -0.99999)
            return Quat(makePerp(), Math::PI);
        else
            return Quat(cross(v).norm(), Math::acos(norm().dot(v.norm())));
    }
    
    Vec3 Vec3::makePerp() const
    {
        if (dot(Vec3::UNIT_Y) < 0.75)
            return cross(Vec3::UNIT_Y);
        else
            return cross(Vec3::UNIT_Z);
    }
    
    Vec3 Vec3::lerp(const Vec3& v, float i) const
    {
        return Vec3((v.x - x) * i + x,
                    (v.y - y) * i + y,
                    (v.z - z) * i + z);
    }
    
    bool Vec3::equals(const Vec3& v, float eps) const
    {
        return Math::equal(x,v.x,eps) && Math::equal(y,v.y,eps) && Math::equal(z,v.z,eps);
    }
    
    std::ostream& operator<<(std::ostream& os, const Vec3& v)
    {
        os << "<Vec3 " << v.x << " " << v.y << " " << v.z << ">";
        return os;
    }
    
    std::string Vec3::toString()
    {
        std::ostringstream outs;
        outs << "Vec3(" << x << ", " << y << ", " << z << ")";
        return outs.str();
    }
    
    bool Vec3::pyEquals( const Vec3& v )
    {
        return equals(v);
    }
    
    boost::python::object Vec3::defineClass()
    {
        return boost::python::class_<Vec3>("Vec3",boost::python::init<>())
            .def(boost::python::init<float,float,float>())
            .def(boost::python::init<Quat,float>())
            .def_readwrite("x", &Vec3::x)
            .def_readwrite("y", &Vec3::y)
            .def_readwrite("z", &Vec3::z)
            .def("__repr__", &Vec3::toString)
            .def("__eq__", &Vec3::pyEquals)
            .def(boost::python::self + boost::python::other<Vec3>())
            .def(boost::python::self - boost::python::other<Vec3>())
            .def(boost::python::self * float())
            .def(boost::python::self / float())
            .def("length",&Vec3::length)
            .def("norm",&Vec3::norm)
            .def("dot",&Vec3::dot)
            .def("cross",&Vec3::cross)
            .def("project",&Vec3::project)
            .def("lerp",&Vec3::lerp)
            .def("getRotTo",&Vec3::getRotTo)
        ;
    }
}

File: acce3/core/vec3.py
from util import feq
import math

class Vec3:
  def __init__(self, x, y, z):
    self.x = float(x)
    self.y = float(y)
    self.z = float(z)
    
  def __repr__(self):
    return 'Vec3(%f, %f, %f)' % (self.x, self.y, self.z)
    
  def __eq__(self, v):
    return feq(self.x, v.x) and feq(self.y, v.y) and feq(self.z, v.z)
    
  def __getitem__(self, i):
    if i == 0:
      return self.x
    elif i == 1:
      return self.y
    elif i == 2:
      return self.z
    else:
      raise KeyError(i)
      
  def __setitem__(self, i, v):
    if i == 0:
      self.x = v
    elif i == 1:
      self.y = v
    elif i == 2:
      self.z = v
    else:
      raise KeyError(i)
    
  def __neg__(self):
    return Vec3(-self.x, -self.y, -self.z)
    
  def __add__(self, v):
    return Vec3(self.x + v.x, self.y + v.y, self.z + v.z)
    
  def __sub__(self, v):
    return Vec3(self.x - v.x, self.y - v.y, self.z - v.z)
    
  def __mul__(self, v):
    if isinstance(v, Vec3):
      return Vec3(self.x * v.x, self.y * v.y, self.z * v.z)
    else:
      return Vec3(self.x * v, self.y * v, self.z * v)
      
  def __div__(self, v):
    if isinstance(v, Vec3):
      return Vec3(self.x / v.x, self.y / v.y, self.z / v.z)
    else:
      return Vec3(self.x / v, self.y / v, self.z / v)
      
  def sq_length(self):
    return self.x * self.x + self.y * self.y + self.z * self.z
    
  def length(self):
    return math.sqrt(self.sq_length())
    
  def norm(self):
    return self / self.length()
    
  def dot(self, v):
    return self.x * v.x + self.y * v.y + self.z * v.z
    
  def lerp(self, v, i):
    return self + ( v - self ) * i
    
  @staticmethod
  def zero():
    return Vec3(0, 0, 0)
    
  @staticmethod
  def test():
    assert Vec3(1, 2, 3) + Vec3(3, 4, 5) == Vec3(4, 6, 8)
    assert Vec3(1, 2, 3) - Vec3(3, 4, 5) == Vec3(-2, -2, -2)
    assert Vec3(1, 2, 3) * Vec3(3, 4, 5) == Vec3(3, 8, 15)
    assert Vec3(1, 2, 3) * 3 == Vec3(3, 6, 9)
    assert Vec3(10, 15, 20) / Vec3(2, 5, 4) == Vec3(5, 3, 5)
    assert Vec3(10, 15, 20) / 5 == Vec3(2, 3, 4)
    assert feq(Vec3(3, 4, 5).sq_length(), 50)
    assert feq(Vec3(3, 4, 5).length(), math.sqrt(50))
    assert Vec3(0, -10, 0).norm() == Vec3(0, -1, 0)
    assert feq(Vec3(1, 2, 3).dot(Vec3(3, 4, 5)), 26)
    assert Vec3(0, 0, 0).lerp(Vec3(10, 20, -30), .5) == Vec3(5, 10, -15)
    
Vec3.test()

